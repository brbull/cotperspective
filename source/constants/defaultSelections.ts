
// Values to be used for index pages, or tests as the default page selection
export const selectedExchange = 'CHICAGO MERCANTILE EXCHANGE'
export const selectedMarket = 'EURO FX'
export const selectedTraderCategory = 'Noncommercial'
