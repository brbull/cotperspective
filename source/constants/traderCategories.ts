export const traderCategories = ['Commercial', 'Noncommercial'] as const
