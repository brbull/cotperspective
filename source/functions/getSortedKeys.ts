
/**
 * Returns a sorted array containing the keys of the given Record/Object
 */
export const getSortedKeys = (data: Record<string, unknown>): string[] => Object.keys(data).sort()
