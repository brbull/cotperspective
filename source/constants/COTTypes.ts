import { COTPropertiesToSelect } from '../constants/COTPropertiesToSelect'
import { traderCategories } from './traderCategories'

export type TraderCategory = typeof traderCategories[number]

export type CSVData = Record<typeof COTPropertiesToSelect[number], string>

export interface FormattedCSVData extends CSVData {
  market: string,
  exchange: string
}

export type MarketsData = Record<string, FormattedCSVData[]>

export type COTData = Record<string, MarketsData>
