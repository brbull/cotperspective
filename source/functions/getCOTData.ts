import unzipper from 'unzipper'
import request from 'request'
import { processData } from './processData'
import { COTData } from '../constants/COTTypes'

const zipBaseURL = 'https://www.cftc.gov/sites/default/files/files/dea/history/'

/**
 * Fetch historical data from CFTC website, and parse to JSON.
 * @param year The latest year of which to load historical data
 * @param minimumEntries Minimum entries the current year should have, otherwise load more data from previous year
 */
export const getCOTData = async (year: number, minimumEntries: number): Promise<COTData> => {
  console.log(`-- fetching COT data for year: ${year}`)
  const zipURL = `${zipBaseURL}deahistfo${year}.zip`
  // Current types for unzipper seem to be incorrect
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const directory = await unzipper.Open.url(request as any, zipURL)
  const file = directory.files.find(d => d.path === 'annualof.txt')
  const contentBuffer = await file?.buffer()
  if (!contentBuffer) throw new Error('CSV content buffer undefined')
  const data = processData(contentBuffer.toString())
  // If not enough entries in current year, then load previous year, and join the data for each market
  if (data['CHICAGO MERCANTILE EXCHANGE']['EURO FX']?.length < minimumEntries) {
    console.log('-- fetching COT data for previous year')
    const previousYearData = await getCOTData(year - 1, minimumEntries)
    for (const exchange in data) {
      for (const market in data[exchange]) {
        const previousData = previousYearData[exchange][market]
        if (previousData) data[exchange][market].push(...previousData)
      }
    }
  }
  return data
}
