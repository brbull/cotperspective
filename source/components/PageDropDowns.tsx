import React, { ReactNode } from 'react'
import { COTData } from '../constants/COTTypes'
import { traderCategories } from '../constants/traderCategories'
import { getPagePath } from '../functions/getFileName'

interface SelectProps {
  children: ReactNode
  defaultValue: string
}

const Select = ({ children, defaultValue }: SelectProps) => (
  <select
    defaultValue={defaultValue}
    className='custom-select form-control js-page-dropdown-select'
  >
    {children}
  </select>
)

export interface PageDropDownsProps {
  data: COTData,
  markets: string[],
  exchanges: string[],
  traderCategories: typeof traderCategories,
  selectedExchange: string,
  selectedMarket: string,
  selectedTraderCategory: string
}

declare global {
  interface Window {
    // Separate namespace to facilitate testing (using mocks)
    cotperspective: {
      assignLocation: typeof window.location.assign
    }
  }
}

/**
 * Handles page navigation after a dropdown option is selected.
 * Note: Might be refactored in the future if the site needs to
 * run React for more complex components, but at the moment this
 * fulfills the requirement while being a statically rendered site.
 */
export const pageSelectScript = `
  window.cotperspective = {
    assignLocation: window.location.assign.bind(window.location)
  }
  document.getElementById('cotperspective').addEventListener('change', (event) => {
    if (event.target.className.includes('js-page-dropdown-select')) {
      window.cotperspective.assignLocation(event.target.value + '.html')
    }
  })
`

export const PageDropDowns = ({
  data,
  markets,
  exchanges,
  traderCategories,
  selectedExchange,
  selectedMarket,
  selectedTraderCategory
}: PageDropDownsProps): JSX.Element => {
  const defaultPage = getPagePath({
    selectedExchange,
    selectedMarket,
    selectedTraderCategory
  })
  return (
    <form style={{ marginTop: 20 }}>
      <div className='row'>
        <div className='col'>
          <div className='form-group'>
            <label>Exchange</label>
            <Select defaultValue={defaultPage}>
              {exchanges.map(exchange =>
                <option
                  key={exchange}
                  value={`${getPagePath({
                    selectedExchange: exchange,
                    selectedMarket: Object.keys(data[exchange])[0],
                    selectedTraderCategory
                  })}`}
                >{exchange}</option>
              )}
            </Select>
          </div>
        </div>
        <div className='col'>
          <div className='form-group'>
            <label>Market</label>
            <Select defaultValue={defaultPage}>
              {markets.map(market =>
                <option
                  key={market}
                  value={`${getPagePath({
                    selectedExchange: selectedExchange,
                    selectedMarket: market,
                    selectedTraderCategory
                  })}`}
                >{market}</option>
              )}
            </Select>
          </div>
        </div>
        <div className='col'>
          <div className='form-group'>
            <label>Trader</label>
            <Select defaultValue={defaultPage}>
              {traderCategories.map(traderCategory =>
                <option
                  key={traderCategory}
                  value={`${getPagePath({
                    selectedExchange: selectedExchange,
                    selectedMarket: selectedMarket,
                    selectedTraderCategory: traderCategory
                  })}`}
                >{traderCategory}</option>
              )}
            </Select>
          </div>
        </div>
      </div>
      {/* Enable page navigation after selecting a dropdown option */}
      <script dangerouslySetInnerHTML={{ __html: pageSelectScript }}/>
    </form>
  )
}
