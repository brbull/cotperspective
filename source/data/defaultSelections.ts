import { DropDownSelections } from '../components/PageDropDowns'
import { traderCategories } from './traderCategories'

/** Index pages, or test default page selection */
export const defaultSelections: DropDownSelections = {
  exchange: 'CHICAGO MERCANTILE EXCHANGE',
  market: 'EURO FX',
  traderCategory: traderCategories[1]
}
