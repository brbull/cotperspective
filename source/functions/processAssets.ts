import CleanCSS from 'clean-css'
import fs from 'fs'
import { buildPath } from '../data/buildPath'
import { execSync } from 'child_process'

const assetsPath = './source/assets'

const copyAssets = (files: string[]) => files.forEach(file =>
  fs.copyFileSync(`${assetsPath}/${file}`, `${buildPath}/${file}`))

export const processAssets = () => {
  console.log('• Copying assets')
  copyAssets(['favicon.ico', 'preview.png'])  
  console.log('• Processing styles')
  const styles = '' + fs.readFileSync('./source/assets/styles.css')
  const minStyles = new CleanCSS().minify(styles).styles
  fs.writeFileSync(`${buildPath}/styles.css`, minStyles)
}
