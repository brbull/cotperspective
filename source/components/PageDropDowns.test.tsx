import React from 'react'
import { PageDropDowns, pageSelectScript } from './PageDropDowns'
import { testData } from '../functions/testUtils/testData'
import { getSortedKeys } from '../functions/getSortedKeys'
import { traderCategories } from '../constants/traderCategories'
import { selectedExchange, selectedMarket, selectedTraderCategory } from '../constants/defaultSelections'
import { shallow, mount } from 'enzyme'
import { JSDOM } from 'jsdom'

const renderPageDropDowns = () => <PageDropDowns
  data={testData}
  exchanges={getSortedKeys(testData)}
  markets={getSortedKeys(testData[selectedExchange])}
  selectedExchange={selectedExchange}
  selectedMarket={selectedMarket}
  selectedTraderCategory={selectedTraderCategory}
  traderCategories={traderCategories} />

/**
 * Verify correct PageDropDowns rendering of test data
 */
test('PageDropDowns snapshot', () => {
  expect(shallow(renderPageDropDowns())).toMatchSnapshot()
})

/**
 * Verify correct PageDropDowns navigation to selected page
 */
test('PageDropDowns navigation', () => {
  // Component's <script> tag does not execute.
  // A new JSDOM instance is created to allow script execution
  const { window } = new JSDOM(
    `<body>
      <div id='cotperspective'></div>
      <script>${pageSelectScript}</script>
    </body>`,
    {
      runScripts: 'dangerously',
      url: 'https://cotperspective.com'
    }
  )

  mount(renderPageDropDowns(), {
    attachTo: window.document.getElementById('cotperspective')
  })

  window.cotperspective.assignLocation = jest.fn()

  const traderCategorySelect = window.document.getElementsByClassName('js-page-dropdown-select')[2] as HTMLSelectElement
  traderCategorySelect.value = 'chicagomercantileexchangeeurofxcommercial'
  traderCategorySelect.dispatchEvent(new window.Event('change', { bubbles: true }))

  expect(window.cotperspective.assignLocation)
    .toHaveBeenLastCalledWith(traderCategorySelect.value + '.html')
})
