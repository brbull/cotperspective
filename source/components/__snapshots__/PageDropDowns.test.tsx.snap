// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageDropDowns snapshot 1`] = `
<form
  style={
    Object {
      "marginTop": 20,
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label>
          Exchange
        </label>
        <Select
          defaultValue="chicagomercantileexchange3monthsofrnoncommercial"
        >
          <option
            key="CHICAGO MERCANTILE EXCHANGE"
            value="chicagomercantileexchange3monthsofrnoncommercial"
          >
            CHICAGO MERCANTILE EXCHANGE
          </option>
          <option
            key="ICE FUTURES U.S."
            value="icefuturesususdollarindexnoncommercial"
          >
            ICE FUTURES U.S.
          </option>
        </Select>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label>
          Market
        </label>
        <Select
          defaultValue="chicagomercantileexchangeeurofxnoncommercial"
        >
          <option
            key="3-MONTH SOFR"
            value="chicagomercantileexchange3monthsofrnoncommercial"
          >
            3-MONTH SOFR
          </option>
          <option
            key="EURO FX"
            value="chicagomercantileexchangeeurofxnoncommercial"
          >
            EURO FX
          </option>
          <option
            key="EURO FX/BRITISH POUND XRATE"
            value="chicagomercantileexchangeeurofxbritishpoundxratenoncommercial"
          >
            EURO FX/BRITISH POUND XRATE
          </option>
        </Select>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="form-group"
      >
        <label>
          Trader
        </label>
        <Select
          defaultValue="chicagomercantileexchangeeurofxnoncommercial"
        >
          <option
            key="Commercial"
            value="chicagomercantileexchangeeurofxcommercial"
          >
            Commercial
          </option>
          <option
            key="Noncommercial"
            value="chicagomercantileexchangeeurofxnoncommercial"
          >
            Noncommercial
          </option>
        </Select>
      </div>
    </div>
  </div>
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "
  window.cotperspective = {
    assignLocation: window.location.assign.bind(window.location)
  }
  document.getElementById('cotperspective').addEventListener('change', (event) => {
    if (event.target.className.includes('js-page-dropdown-select')) {
      window.cotperspective.assignLocation(event.target.value + '.html')
    }
  })
",
      }
    }
  />
</form>
`;
